Feature access lists

Qwiko opened the pull request:
I have worked on adding access-list generation support from the settings-repository by using [Aerleon](https://aerleon.readthedocs.io/en/latest/).

The idea behind the implementation is to reuse as much as possible from aerleon.
Like validating the syntax is not super simple so I thought just render the access-lists directly as validation and handle the errors.

The core functionality is there but is a rather large change that needs to be discussed futher on the final implementation if it is a feature wanted in NMS?

@indy-independence @YotaYota please provide feedback.

Proposed usage:

<details>
  <summary>global/access-lists.yaml</summary>

```yaml
---
network_definitions:
  "ONEZEROZEROONE":
    - address: 1.0.0.1
    - address: 2606:4700:4700::1001
  "ONEONEONEONE":
    - address: 1.1.1.1
    - address: 2606:4700:4700::1111
service_definitions:
  "DHCP":
    - port: 67
      protocol: udp
access_lists:
  "COUNTERS":
    comment: |
      Import this at the start of a access_list to enable counters on eos or nxos
    include_only: true
    terms:
      - verbatim:
          eos: " counters per-entry"
          nxos: " statistics per-entry"
        platform:
          - eos
          - nxos
  "INCLUDED-ACL":
    comment: | 
      This is an access-list that is included in other lists
    include_only: true
    terms:
      - destination-address: "ONEZEROZEROONE"
        action: "accept"
  "ACL-TEST":
    comment: Some acl
    header_map:
      ios: "{ACL_NAME} {INET_FAMILY} noverbose"
    inet_families:
      - ipv4
      - ipv6
    terms:
      - include: COUNTERS

      - include: INCLUDED-ACL

      - name: "accept dhcp"
        destination-address: ONEONEONEONE
        destination-port: DHCP
        protocol: "udp"
        action: "accept"
```
</details>

Produces the following access-lists ->

<details>
  <summary>ios</summary>

`noverbose` for ios removes all comments
```
! $Id:$
! $Date:$
! $Revision:$
no ipv6 access-list ACL-TEST
ipv6 access-list ACL-TEST


permit ipv6 any host 2606:4700:4700::1001


permit udp any host 2606:4700:4700::1111 eq 67

exit

no ip access-list extended ACL-TEST
ip access-list extended ACL-TEST


permit ip any host 1.0.0.1


permit udp any host 1.1.1.1 eq 67

exit
```
</details>
<details>
  <summary>eos</summary>

```
! $Id:$
! $Date:$
! $Revision:$
no ip access-list ACL-TEST
ip access-list ACL-TEST
remark $Id:$
remark Some acl


remark 9e0716_term_1
counters per-entry


remark 9daf9d_term_1
permit ip any host 1.0.0.1


remark 75eec4_accept_dhcp
permit udp any host 1.1.1.1 eq bootps

exit

no ipv6 access-list ACL-TEST
ipv6 access-list ACL-TEST
remark $Id:$
remark Some acl


remark 9e0716_term_1
counters per-entry


remark 9daf9d_term_1
permit ipv6 any host 2606:4700:4700::1001


remark 75eec4_accept_dhcp
permit udp any host 2606:4700:4700::1111 eq bootps

exit
```
</details>
<details>
  <summary>junos</summary>

```
firewall {
    family inet {
        /*
        ** $Id:$
        ** $Date:$
        ** $Revision:$
        **
        ** Some acl
        */
        replace: filter ACL-TEST {
            interface-specific;
            term 9daf9d_term_1 {
                from {
                    destination-address {
                        1.0.0.1/32;
                    }
                }
                then accept;
            }
            term 75eec4_accept_dhcp {
                from {
                    destination-address {
                        1.1.1.1/32;
                    }
                    protocol udp;
                    destination-port 67;
                }
                then accept;
            }
        }
    }
}
firewall {
    family inet6 {
        /*
        ** $Id:$
        ** $Date:$
        ** $Revision:$
        **
        ** Some acl
        */
        replace: filter ACL-TEST {
            interface-specific;
            term 9daf9d_term_1 {
                from {
                    destination-address {
                        2606:4700:4700::1001/128;
                    }
                }
                then accept;
            }
            term 75eec4_accept_dhcp {
                from {
                    destination-address {
                        2606:4700:4700::1111/128;
                    }
                    next-header udp;
                    destination-port 67;
                }
                then accept;
            }
        }
    }
}
```
</details>

Qwiko committed:
Initial commit
Added aerleon access-list support in settings

Qwiko committed:
Updated documentation

Qwiko committed:
Check if include_acl is not None

Qwiko committed:
Optimize hashing for redis to make smaller keys and can handle objects with __dict__, like the Naming object.
Change redis-caching so different devices with the same access-lists will not generate twice.

Qwiko committed:
Compute hash only from the individual term + index information so it wont change all the time when other terms change or is added.

Qwiko committed:
Add AccessListGenerationError exception handling

Qwiko committed:
Use random ips instead of hardcoded ones

Qwiko committed:
Mark as safe

Qwiko committed:
Use \w
Use another var-name instead of builtin hash
Iterate over .values()

Qwiko committed:
Remove TODO

Qwiko committed:
Remove unused vars

indy-independence commented on src/cnaas_nms/db/settings_fields.py:
```diff
@@ -423,6 +423,111 @@ class f_port_template(BaseModel):
+                },
+                sort_keys=True,
+            )
+            return hashlib.sha256(hash_input.encode()).hexdigest()[:6]
```
will this code run a lot of times? maybe it could use a faster hash like crc32 or something if we throw away most of the data anyhow?

Qwiko replied:
I think I will rework how that is handled. Maybe just remove it altogether and force the user to format the name correctly.

Qwiko committed:
Use list instead of set, unique and sort list
Change behavior to limit infinite recursion

Qwiko committed:
Import TypedDict from typing_extension for python < 3.12 to support aerleon

Qwiko committed:
Change wording

Qwiko committed:
Simplify and added additional validation for access_lists.

Qwiko committed:
Added more tests

Qwiko committed:
Fix for mypy that include_acl is a string

Qwiko committed:
Add support for optional files in directory structure

Qwiko committed:
Added documentation for access-lists
Changed headers
Added Table of Contents

Qwiko committed:
Remove trailing whitespace

Qwiko committed:
Instead of raising an error for an incompatible platform return nothing.

Qwiko committed:
Revert pydantic typing issue resolved in aerleon 1.12.1

Qwiko committed:
Added system_access_lists setting

Qwiko committed:
access_lists,yml is now a global-only optional file
Simplified redis-key hash
Marked acl-related functions as internal
Added dynamic lookup from vxlans and interface to generate access-lists for devices

Qwiko committed:
Mark IPs as safe

Qwiko committed:
Change to assert

Qwiko commented:
Waiting on PR: https://github.com/aerleon/aerleon/pull/443 so nested includes works through the aerleon api.

Qwiko committed:
Remove assertion

Qwiko committed:
Remove unused variable

Qwiko committed:
Remove acl-related keys from SETTINGS_KEYS_TO_MERGE

Qwiko committed:
Remove from docs

Qwiko committed:
Remove system_access_lists from device vars

Qwiko committed:
Upgrade pyyaml to 6.0.3

Qwiko committed:
Remove random ip functions from tests

Qwiko committed:
Change variable names and added more comments

Qwiko committed:
Moved system_access_lists verification to get_settings.
Added access_list verification to settings/model api
Updated tests

Qwiko committed:
Remove comment

Qwiko committed:
Move acl generation in populate_device_vars so it will work during device_init for DIST-switches.

Qwiko committed:
Add system_access_lists to default SETTINGS_KEYS_TO_MERGE

Qwiko committed:
Remove include_only parameter

Qwiko committed:
Update documentation

Qwiko committed:
Update aerleon to 1.13.0

Qwiko committed:
Update to new aerleon typeddict and import syntax.

Qwiko commented:
I feel like this PR is ready for review now.

Whats done?

- Added integration with aerleon for access_lists generation support from the settings-repo.
- Exposes fully generated access_lists as device_variables usable in jinja-templates.
Returns a dict called `access_lists` as: `{access_list_name: "access_list_as_text"}`
- Added a new file access_lists.yml that can only exist on the global-folder level.
This file is optional and does not need to be setup in the repository making this functionality opt-in.
- Access lists are dynamically fetched from settings that is set on device-interfaces or vlan-interfaces. (vlan-interface only for dist-switches).
- Added a base_system option: system_access_lists to make other device types generate access_lists.
Example usage for system_access_lists is SSH/SNMP or API access-lists.
This setting is by default merge so global level + primary group + device-specific options are merged.
Making it possible having global access_access lists defined once and adding any specific ones for a group or device.
- Updated api-test functionality to include acl-errors making it easier to build access lists in for example vs-code.
- Added tests.
- Updated documentation.

Left to do
- Fully test in lab-environment

This PR should be squash-merged to make it more clean in the git-history.

Qwiko committed:
Remove access_lists.yml from docs for groups and devices.

Qwiko committed:
Add documentation how to use established option on tcp and udp.

indy-independence commented on src/cnaas_nms/api/device.py:
```diff
@@ -385,6 +386,11 @@ def put(self, device_id: int):
+                    msg = str(e)
+                    logger.error(msg)
+                    session.rollback()
+                    return empty_result(status="error", data=msg)
```
exceptions above return http 500, should it be the same here?

Qwiko replied:
Good catch! Yes it should be the same.

Qwiko committed:
Return error 500

Qwiko committed:
Add  CaptuireHandler that can save records in a list-buffer

Qwiko committed:
Override aerleon logs and emit logs over cnaas logger instead to display warning messages to the user
